libs/langchain-google-cloud-sql-pg/src/vectorStore.ts

```diff
@@ -0,0 +1,423 @@
+    )
+  }
+
+  static async fromTexts(texts: string[], metadatas: object[] | object, embeddings: EmbeddingsInterface, dbConfig: dbConfigArgs): Promise<VectorStore> {
+    const documents: Document[] = [];
+
+    for (let i = 0; i < texts.length; i += 1) {
+      const doc = new Document({
+        pageContent: texts[i],
+        metadata: Array.isArray(metadatas) ? metadatas[i] : metadatas
+      })
+      documents.push(doc);
+    }
+
+    return PostgresVectorStore.fromDocuments(documents, embeddings, dbConfig);
+  }
+
+  static async fromDocuments(docs: Document[], embeddings: EmbeddingsInterface, dbConfig: dbConfigArgs): Promise<VectorStore> {
+    const { engine } = dbConfig;
+    const { tableName } = dbConfig;
+    const config = dbConfig.dbConfig;
+    const vectorStore = await this.create(engine, embeddings, tableName, config);
+
+    await vectorStore.addDocuments(docs)
+
+    return vectorStore;
+  }
+
+  async addVectors(vectors: number[][], documents: Document[], options?: { ids?: string[] }): Promise<string[] | void> {
+    let ids: string[] = [];
+    const metadatas: Record[] = []
+
+    if (vectors.length !== documents.length) {
+      throw new Error("The number of vectors must match the number of documents provided.");
+    }
+
+    if (options?.ids) {
+      ids = options.ids;
+    } else {
+      documents.forEach(document => {
+        if (document.id !== undefined) {
+          ids.push(document.id);
+        } else {
+          ids.push(uuidv4());
+        }
+      });
+    }
+
+    if (options && options.ids && options.ids.length !== documents.length) {
+      throw new Error("The number of ids must match the number of documents provided.");
+    }
+
+    documents.forEach(document => {
+      metadatas.push(document.metadata)
+    });
+
+    const tuples = customZip(ids, documents, vectors, metadatas);
+
+    // Insert embeddings
+    for (const [id, document, embedding, metadata] of tuples) {
+      const metadataColNames = this.metadataColumns.length > 0 ? `, "${this.metadataColumns.join("\",\"")}"` : "";
+
+      let stmt = `INSERT INTO "${this.schemaName}"."${this.tableName}"("${this.idColumn}", "${this.contentColumn}", "${this.embeddingColumn}" ${metadataColNames}`
+      const values: { [key: string] } = {
+        id,
+        content: document.pageContent,
+        embedding: `[${embedding.toString()}]`
+      }
+      let valuesStmt = " VALUES (:id, :content, :embedding";
+
+      // Add metadata
+      const extra = metadata;
+      for (const metadataColumn of this.metadataColumns) {
+        if (Object.prototype.hasOwnProperty.call(metadata, metadataColumn)) {
+          valuesStmt += `, :${metadataColumn}`;
+          values[metadataColumn] = metadata[metadataColumn]
+          delete extra[metadataColumn]
+        } else {
+          valuesStmt += " ,null"
+        }
+      }
+
+      // Add JSON column and/or close statement
+      stmt += this.metadataJsonColumn ? `, ${this.metadataJsonColumn})` : ")";
```

Comment:
OOC why not just support `metadataJsonColumn` vs spreading metadata into other columns?

Reply:
This pattern follows database best practices for relational databases by using individual columns versus all data in a JSON column. This allows for low latency filtering and type checking.

Reply:
And the JSON column is just to make onboarding easier?

Reply:
Yes and we don't want users to lose metadata if not specifically defined